### PR TITLE
Add bind_addr for udp stream

### PR DESCRIPTION
**Pull request by hainesc**
The bind_addr is missing recently, see https://github.com/hickory-dns/hickory-dns/issues/2190

### CODE REVIEW

**Comment by hainesc** [comment]
ping @bluejekyll @djc, may you take a look? Thanks very much!

File: crates/proto/src/udp/udp_stream.rs [hunk 1]
```diff
@@ -264,12 +264,19 @@ impl<S: UdpSocket + 'static> NextRandomUdpSocket<S> {
 
 impl<S: DnsUdpSocket> NextRandomUdpSocket<S> {
     /// Create a future with generator
-    pub(crate) fn new_with_closure(name_server: &SocketAddr, func: UdpCreator<S>) -> Self {
-        let bind_address = match *name_server {
-            SocketAddr::V4(..) => SocketAddr::new(IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)), 0),
-            SocketAddr::V6(..) => {
-                SocketAddr::new(IpAddr::V6(Ipv6Addr::new(0, 0, 0, 0, 0, 0, 0, 0)), 0)
-            }
+    pub(crate) fn new_with_closure(
```
Comment (hainesc) [inline crates/proto/src/udp/udp_stream.rs]: Here I can add a new function to keep compatible

Comment (hainesc) [reply crates/proto/src/udp/udp_stream.rs]: Done

**Comment by djc** [comment]
> ping @bluejekyll @djc, may you take a look? Thanks very much!

There's no need to ping us immediately after filing a PR.

**Comment by djc** [comment]
@bluejekyll see discussion in #2190 on whether this is the right solution.

File: crates/proto/src/udp/udp_stream.rs [hunk 2]
```diff
@@ -264,12 +264,25 @@ impl<S: UdpSocket + 'static> NextRandomUdpSocket<S> {
...
+    }
+
+    /// Create a future with generator
+    pub(crate) fn new_with_bind_addr_and_closure(
+        name_server: &SocketAddr,
+        bind_addr: &Option<SocketAddr>,
+        func: UdpCreator<S>,
+    ) -> Self {
+        let bind_address = match bind_addr {
+            Some(ba) => *ba,
```
[the remark this answers is not in the document]
Comment (hainesc) [reply crates/proto/src/udp/udp_stream.rs:279]: Good point, I will take a look what other(such as std lib) do on this problem.

**Comment by bluejekyll** [comment]
> @bluejekyll see discussion in #2190 on whether this is the right solution.

That's a good discussion, let's finish that out.

File: crates/proto/src/udp/udp_client_stream.rs [hunk 3]
```diff
@@ -36,6 +36,7 @@ where
     MF: MessageFinalizer,
 {
     name_server: SocketAddr,
+    bind_addr: Option<SocketAddr>,
```
Comment (bluejekyll) [inline crates/proto/src/udp/udp_client_stream.rs:39]: Based on the discussion in #2190, this should probably be IPaddr instead. I do think we should still consider how we want to deal with IPv4 vs IPv6...

**Pull request closed without merging**